File: src/main/scala/org/zezutom/learningspark/scala/WordCount.scala
package org.zezutom.learningspark.scala

import org.apache.spark.rdd.RDD

object WordCount {
  def wc(lines:RDD[String]): RDD[(String, Int)] = {
    lines.flatMap(_.split("\\s"))                             // Split on any white character
      .filter(!_.isEmpty)                                     // Only count words as non-empty strings
      .map(_.stripSuffix(",").stripSuffix(".").toLowerCase)   // Remove punctuation and make all words lowercase
      .map(word => (word, 1)).reduceByKey(_ + _).sortBy(_._1)
  }
}

File: src/main/java/org/zezutom/learningspark/WordCount.java
package org.zezutom.learningspark;

import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import scala.Tuple2;

import java.util.Arrays;

public class WordCount {

    public static JavaPairRDD<String, Integer> wc(JavaRDD<String> lines) {
        return lines
                .flatMap(s -> Arrays.asList(s.split("\\s")))        // Split on any white character
                .filter(word -> !word.isEmpty())                    // Only count words as non-empty strings
                .map(word -> word.replaceAll(",", "")
                        .replaceAll("\\.", "").toLowerCase())         // Remove punctuation and make all words lowercase
                .mapToPair(s -> new Tuple2<>(s, 1))
                .reduceByKey((i1, i2) -> i1 + i2)
                .sortByKey();
    }
}
